### PR TITLE
drivers: eeprom: add EEPROM support for STM32L1 series

### PR DESCRIPTION
**Pull request by KwonTae-young**
Add EEPROM support for STM32.
Tested on `96b_wistrio` board.

Currently only the `STM32L1` series is supported.

**tests/drivers/eeprom**
```
*** Booting Zephyr OS version 2.1.0  ***
Running test suite eeprom_api
===================================================================
starting test - test_size
PASS - test_size
===================================================================
starting test - test_out_of_bounds
W: attempt to write past device boundary
PASS - test_out_of_bounds
===================================================================                   
starting test - test_write_and_verify                                                 
PASS - test_write_and_verify                                                          
===================================================================                   
starting test - test_zero_length_write                                                
PASS - test_zero_length_write                                                         
===================================================================                   
Test suite eeprom_api succeeded                                                       
===================================================================                   
PROJECT EXECUTION SUCCESSFUL 
```

Please let me know if there is anything strange. :cold_sweat:

### CODE REVIEW

File: dts/bindings/mtd/st,stm32-eeprom.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Copyright (c) 2019 Kwon Tae-young <tykwon@m2i.co.kr>
+# SPDX-License-Identifier: Apache-2.0
+
+include: eeprom-base.yaml
+
+title: STM32 EEPROM devices
+
+description: |
```
Comment (erwango) [inline dts/bindings/mtd/st,stm32-eeprom.yaml]: Nitpick, no need to split the description on 2 lines here.

File: dts/bindings/mtd/st,stm32-eeprom.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    STM32 compatible EEPROM
```
Comment (henrikbrixandersen) [inline dts/bindings/mtd/st,stm32-eeprom.yaml]: Perhaps mention that is the on-chip EEPROM here?

**Review by henrikbrixandersen — APPROVED** [review]
Thank you for the driver submission! Looks good to me.

**Review by erwango — CHANGES_REQUESTED** [review]
> Please let me know if there is anything strange.

What is strange is that I only have one tiny comment :-)
Thanks for this good work!

One question, do you want to maintain this driver (ie contribute to code reviews, bug fixes and evolutions)?
If yes, please add your name in codeowner file.

File: dts/bindings/mtd/st,stm32-eeprom.yaml (hunk 1 above)
Comment (KwonTae-young) [reply dts/bindings/mtd/st,stm32-eeprom.yaml]: Done

File: dts/bindings/mtd/st,stm32-eeprom.yaml (hunk 2 above)
Comment (KwonTae-young) [reply dts/bindings/mtd/st,stm32-eeprom.yaml]: Done

**Comment by KwonTae-young** [comment]
@erwango 
> If yes, please add your name in codeowner file.

Yes. I put a name in the `CODEOWNER` file.
```
/drivers/eeprom/eeprom_stm32.c            @KwonTae-young
```

**Review by henrikbrixandersen — APPROVED** [review]

**Review by erwango — CHANGES_REQUESTED** [review]
@KwonTae-young, thanks for update:

Late comment, I don't see a part where you link EEPROM activation to EEPROM_STM32.
Something like follows(in `soc/arm/st_stm32/common/Kconfig.defconfig.series`):
```
if I2C

config I2C_STM32
	default y

endif
```

Also, maybe you can update `tests/drivers/eeprom` to have the test done automatically done by CI.

**Review by dbkinder — APPROVED** [review]
no issues with the doc changes, thanks!

**Comment by KwonTae-young** [comment]
@erwango
I added the following to `soc/arm/st_stm32/common/Kconfig.defconfig.series`.
```
if EEPROM

config EEPROM_STM32
	default y

endif # EEPROM
```

[Checking the CI log](https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/57947/summary/console), `tests/drivers/eeprom` are not tested on `96b_wistrio`.
Should I modify `testcase.yaml` to be tested on `96b_wistrio`?
This is my first time using `tests/`.  :sweat_smile:
``` yaml
tests:
  peripheral.eeprom:
    platform_whitelist: native_posix native_posix_64
    tags: drivers userspace
  peripheral.eeprom.96b_wistrio:
    platform_whitelist: 96b_wistrio
    tags: drivers userspace
```
```
root:zephyr# ./scripts/sanitycheck -T tests/drivers/eeprom/ -v
Renaming output directory to /work/zephyrproject_eeprom/zephyr/sanity-out.20
JOBS: 4
Reading test case configuration files under /work/zephyrproject_eeprom/zephyr/tests/drivers/eeprom...
Reading platform configuration files under /work/zephyrproject_eeprom/zephyr/boards...
Reading platform configuration files under /work/zephyrproject_eeprom/zephyr/scripts/sanity_chk/boards...
Selecting default platforms per test case
Building initial testcase list...
2 test configurations selected, 423 configurations discarded due to filters.
Adding tasks to the queue...
1/2 native_posix              tests/drivers/eeprom/peripheral.eeprom             PASSED (native 0.070s)
2/2 96b_wistrio               tests/drivers/eeprom/peripheral.eeprom.96b_wistrio PASSED (build)

2 of 2 tests passed (100.00%), 0 failed, 0 skipped with 0 warnings in 16.71 seconds
In total 2 test cases were executed on 2 out of total 213 platforms (0.94%)
root:zephyr#
```

File: drivers/eeprom/eeprom_stm32.c [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+	if ((offset + len) > config->size) {
+		LOG_WRN("attempt to read past device boundary");
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&lock, K_FOREVER);
+
+	while (len) {
+		*pbuf = *(__IO uint32_t*)(config->addr + offset);
```
Comment (Mani-Sadhasivam) [inline drivers/eeprom/eeprom_stm32.c]: Why are you casting to `uint32_t` pointer? Shouldn't it be `u8_t` instead?

```
*pbuf = *(__IO u8_t*)(config->addr + offset);
```

**Review by Mani-Sadhasivam — CHANGES_REQUESTED** [review]
Just a minor comment. Otherwise looks good.

File: drivers/eeprom/eeprom_stm32.c (hunk 3 above)
Comment (KwonTae-young) [reply drivers/eeprom/eeprom_stm32.c]: Done

**Review by Mani-Sadhasivam — APPROVED** [review]

**Comment by erwango** [comment]
@KwonTae-young, following is good indeed:
```
  peripheral.eeprom.96b_wistrio:
    platform_whitelist: 96b_wistrio
    tags: drivers userspace
```
Though, it would be nice to have something more generic, so we don't have to add each board systematically, for instance
```
  peripheral.eeprom.generic:
    depends_on: eeprom
    tags: drivers userspace
```
Then, when a new board is updated to support eeprom, we just need to add `eeprom` in `supported` category of `board.yaml` file

**Comment by KwonTae-young** [comment]
@erwango
Added testcase for boards that support EEPROM.
Below is the [CI log](https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/58080/2/console).
```
2019-12-12 01:53:39: 127/738 96b_wistrio               tests/drivers/eeprom/peripheral.eeprom.generic     PASSED (build)
```
<br/>

The following is a test log assuming that EEPROM is added to another board.
``` diff
diff --git a/boards/arm/stm32f4_disco/stm32f4_disco.yaml b/boards/arm/stm32f4_disco/stm32f4_disco.yaml
index a3c30933b1..d3cbfb2992 100644
--- a/boards/arm/stm32f4_disco/stm32f4_disco.yaml
+++ b/boards/arm/stm32f4_disco/stm32f4_disco.yaml
@@ -12,3 +12,4 @@ supported:
   - pwm
   - counter
   - usb
+  - eeprom
```
``` console
root:zephyr# ./scripts/sanitycheck -T tests/drivers/eeprom/ -v
Renaming output directory to /work/zephyrproject_eeprom/zephyr/sanity-out.31
JOBS: 4
Reading test case configuration files under /work/zephyrproject_eeprom/zephyr/tests/drivers/eeprom...
Reading platform configuration files under /work/zephyrproject_eeprom/zephyr/boards...
Reading platform configuration files under /work/zephyrproject_eeprom/zephyr/scripts/sanity_chk/boards...
Selecting default platforms per test case
Building initial testcase list...
3 test configurations selected, 421 configurations discarded due to filters.
Adding tasks to the queue...
1/3 stm32f4_disco             tests/drivers/eeprom/peripheral.eeprom.generic     FAILED Build failure (build)

	see: sanity-out/stm32f4_disco/tests/drivers/eeprom/peripheral.eeprom.generic/build.log
2/3 native_posix              tests/drivers/eeprom/peripheral.eeprom             PASSED (native 0.068s)
3/3 96b_wistrio               tests/drivers/eeprom/peripheral.eeprom.generic     PASSED (build)

2 of 3 tests passed (66.67%), 1 failed, 0 skipped with 0 warnings in 24.89 seconds
In total 2 test cases were executed on 3 out of total 213 platforms (1.41%)
root:zephyr# 
```

**Review by erwango — APPROVED** [review]

**Pull request merged**